lib/auth0/auth0.js

```diff
@@ -0,0 +1,207 @@
+        if (req.method === 'GET') {
+            if (error.redirect) {
+                res.writeHead(302, { Location: error.redirect });
+                return res.end(JSON.stringify(error));
+            }
+            res.writeHead(error.code || 401, { 
+                'Content-Type': 'text/html', 
+                'Cache-Control': 'no-cache' 
+            });
+            return res.end(getNotAuthorizedHtml(baseUrl + '/login'));
+        }
+        else {
+            // Reject all other requests
+            return error(error, res);
+        }            
+    };
+    if (typeof options.authorized === 'string') {
+        options.authorized = [ options.authorized ];
+    }
+    if (Array.isArray(options.authorized)) {
+        var authorized = [];
+        options.authorized.forEach(function (a) {
+            authorized.push(a.toLowerCase());
+        });
+        options.authorized = function (ctx, res) {
+            if (ctx.user.email_verified) {
+                for (var i = 0; i < authorized.length; i++) {
+                    var email = ctx.user.email.toLowerCase();
+                    if (email === authorized[i] || authorized[i][0] === '@' && email.indexOf(authorized[i]) > 1) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+    }
+    if (typeof options.exclude === 'string') {
+        options.exclude = [ options.exclude ];
+    }
+    if (Array.isArray(options.exclude)) {
+        var exclude = options.exclude;
+        options.exclude = function (ctx, res, appPath) {
+            return exclude.indexOf(appPath) > -1;
+        }
+    }
+
+    return createAuthenticatedWebtask(webtask, options);
+};
+
+function createAuthenticatedWebtask(webtask, options) {
+
+    // Inject middleware into the HTTP pipeline before the webtask handler
+    // to implement authentication endpoints and perform authentication 
+    // and authorization.
+
+    return function (ctx, req, res) {
+        if (!req.x_wt.jtn || !req.x_wt.container) {
+            return error({
+                code: 400,
+                message: 'Auth0 authentication can only be used with named webtasks.'
+            }, res);
+        }
+
+        var routingInfo = getRoutingInfo(req);
+        if (!routingInfo) {
+            return error({
+                code: 400,
+                message: 'Error processing request URL path.'
+            }, res);
+        }
+        switch (req.method === 'GET' && routingInfo.appPath) {
+            case '/login': handleLogin(options, ctx, req, res, routingInfo); break;
+            case '/callback': handleCallback(options, ctx, req, res, routingInfo); break;
+            default: handleAppEndpoint(webtask, options, ctx, req, res, routingInfo); break;
+        };
+        return;
+    };
+}
+
+function getRoutingInfo(req) {
+    var routingInfo = url.parse(req.url, true);
+    var segments = routingInfo.pathname.split('/');
```

Comment:
To confirm: There is no path that would lead to there not being a leading `/` or an empty `pathname` here, right?